From 6c15a670fd9979069075bf1665d04b7a5460000d Mon Sep 17 00:00:00 2001
From: Sam Parkinson <sam@sam.today>
Date: Fri, 3 Jun 2016 20:59:02 +1000
Subject: [PATCH] Add even/odd row css nodes for tree view

With css regions, it was possible to stripe the treeview rows with
different background colors.  While this is not a design goal for
GNOME, some downstreams (like Sugar) saw this as a regression.

This commit does the laziest possible fix; simply having a "even"
and "odd" css node, that can be addressed by "row.even" and
"row.odd" via the css.

https://bugzilla.gnome.org/show_bug.cgi?id=757495
---
 gtk/gtktreeview.c              | 45 +++++++++++++++++++++++++++++++++++++++++-
 gtk/theme/Adwaita/_common.scss |  2 ++
 2 files changed, 46 insertions(+), 1 deletion(-)

diff --git a/gtk/gtktreeview.c b/gtk/gtktreeview.c
index 757e072..02a9f09 100644
--- a/gtk/gtktreeview.c
+++ b/gtk/gtktreeview.c
@@ -134,12 +134,18 @@
  * ┊   ┊
  * │   ╰── <column header>
  * │
+ * ├── row.even
+ * ├── row.odd
+ * │
  * ╰── [rubberband]
  * ]|
  *
  * GtkTreeView has a main CSS node with name treeview and style class .view.
  * It has a subnode with name header, which is the parent for all the column
  * header widgets' CSS nodes.
+ * For styling rows, there are the row.even and row.odd nodes.  Note that
+ * there is not a node for every row, only a node for the even/odd style.
+ * The row nodes also do not support to box sizing properties (eg. margin).
  * For rubberband selection, a subnode with name rubberband is used.
  */
 
@@ -314,6 +320,8 @@ struct _GtkTreeViewPrivate
 
   /* CSS nodes */
   GtkCssNode *header_node;
+  GtkCssNode *even_node;
+  GtkCssNode *odd_node;
 
   /* Scroll position state keeping */
   GtkTreeRowReference *top_row;
@@ -900,6 +908,10 @@ static void     remove_scroll_timeout                (GtkTreeView *tree_view);
 
 static void     grab_focus_and_unset_draw_keyfocus   (GtkTreeView *tree_view);
 
+static void gtk_tree_view_row_invalidated (GtkCssNode        *node,
+                                           GtkCssStyleChange *style_change,
+                                           GtkTreeView       *tree_view);
+
 /* Gestures */
 static void gtk_tree_view_column_multipress_gesture_pressed (GtkGestureMultiPress *gesture,
                                                              gint                  n_press,
@@ -1851,6 +1863,22 @@ gtk_tree_view_init (GtkTreeView *tree_view)
   gtk_css_node_set_state (priv->header_node, gtk_css_node_get_state (widget_node));
   g_object_unref (priv->header_node);
 
+  priv->even_node = gtk_css_node_new ();
+  gtk_css_node_set_name (priv->even_node, I_("row"));
+  g_signal_connect (priv->even_node, "style-changed",
+                    G_CALLBACK (gtk_tree_view_row_invalidated), tree_view);
+  gtk_css_node_insert_after (widget_node, priv->even_node, priv->header_node);
+  gtk_css_node_add_class (priv->even_node, g_quark_from_static_string ("even"));
+  gtk_css_node_set_state (priv->even_node, gtk_css_node_get_state (widget_node));
+
+  priv->odd_node = gtk_css_node_new ();
+  gtk_css_node_set_name (priv->odd_node, I_("row"));
+  g_signal_connect (priv->odd_node, "style-changed",
+                    G_CALLBACK (gtk_tree_view_row_invalidated), tree_view);
+  gtk_css_node_insert_after (widget_node, priv->odd_node, priv->even_node);
+  gtk_css_node_add_class (priv->odd_node, g_quark_from_static_string ("odd"));
+  gtk_css_node_set_state (priv->odd_node, gtk_css_node_get_state (widget_node));
+
   priv->multipress_gesture = gtk_gesture_multi_press_new (GTK_WIDGET (tree_view));
   gtk_gesture_single_set_button (GTK_GESTURE_SINGLE (priv->multipress_gesture), 0);
   g_signal_connect (priv->multipress_gesture, "pressed",
@@ -3120,6 +3148,15 @@ get_current_selection_modifiers (GtkWidget *widget,
 }
 
 static void
+gtk_tree_view_row_invalidated (GtkCssNode        *node,
+                               GtkCssStyleChange *change,
+                               GtkTreeView       *tree_view)
+{
+  if (gtk_css_style_change_affects (change, GTK_CSS_AFFECTS_REDRAW))
+    gtk_widget_queue_draw (GTK_WIDGET (tree_view));
+}
+
+static void
 gtk_tree_view_multipress_gesture_pressed (GtkGestureMultiPress *gesture,
                                           gint                  n_press,
                                           gdouble               x,
@@ -5236,8 +5273,13 @@ gtk_tree_view_bin_draw (GtkWidget      *widget,
 						   GTK_RBNODE_FLAG_SET (node, GTK_RBNODE_IS_PARENT),
 						   node->children?TRUE:FALSE);
 
-          gtk_style_context_save (context);
+          if (_gtk_rbtree_node_get_index (tree, node) % 2) {
+            gtk_style_context_save_to_node (context, tree_view->priv->odd_node);
+          } else {
+            gtk_style_context_save_to_node (context, tree_view->priv->even_node);
+          }
 
+          gtk_style_context_save (context);
           state = gtk_cell_renderer_get_state (NULL, widget, flags);
           gtk_style_context_set_state (context, state);
 
@@ -5425,6 +5467,7 @@ gtk_tree_view_bin_draw (GtkWidget      *widget,
 	    }
 
           gtk_style_context_restore (context);
+          gtk_style_context_restore (context);
 	  cell_offset += gtk_tree_view_column_get_width (column);
 	}
 
diff --git a/gtk/theme/Adwaita/_common.scss b/gtk/theme/Adwaita/_common.scss
index 4a8d335..2b877a3 100644
--- a/gtk/theme/Adwaita/_common.scss
+++ b/gtk/theme/Adwaita/_common.scss
@@ -3457,6 +3457,8 @@ list {
 
 row {
   transition: all 150ms $ease-out-quad;
+  color: $text_color;
+  background-color: $base_color;
 
   &:hover { transition: none; }
 
-- 
2.7.4

